ngmt/modules/td/_pham.py

```diff
@@ -0,0 +1,372 @@
+
+        # Check if data has the correct shape
+        if data.shape[1] != 6:
+            raise ValueError("Input data must have 6 columns (3 for accelerometer and 3 for gyro)")
+
+        # Check if sampling frequency is positive
+        if sampling_freq_Hz <= 0:
+            raise ValueError("Sampling frequency must be positive")
+
+        # Check if plot_results is a boolean
+        if not isinstance(plot_results, bool):
+            raise ValueError("plot_results must be a boolean value")
+
+        # Select acceleration data and convert it to numpy array format
+        accel = data.iloc[:, 0:3].copy()
+        accel = accel.to_numpy()
+
+        # Select gyro data and convert it to numpy array format
+        gyro = data.iloc[:, 3:6].copy()
+        gyro = gyro.to_numpy()
+
+        # Check unit of acceleration data if it is in g or m/s^2
+        if acceleration_data_unit == "m/s^2":
+            # Convert acceleration data from m/s^2 to g (if not already is in g)
+            accel /= 9.81
+
+        # Check unit of acceleration data if it is in g or m/s^2
+        if gyro_data_unit == "rad/s":
+            # Convert acceleration data from rad/s to deg/s (if not already is in deg/s)
+            gyro = np.rad2deg(gyro)
+
+        # Convert gyro data unit from deg/s to rad/s
+        gyro *= self.gyro_convert_unit
+
+        # Compute the variance of the moving window of gyro signal
+        gyro_vars = []
+
+        for i in range(3):
+            gyro_var = preprocessing.moving_var(data=gyro[:, i], window=sampling_freq_Hz)
+            gyro_vars.append(gyro_var)
+
+        gyro_var_1, gyro_var_2, gyro_var_3 = gyro_vars
+
+        # Threshold value for identifying periods where the variance is low
+        thr = 2 * 10**-4
+        
+        # Identify periods where the variance is below the threshold
+        self.locs = (gyro_var_1 <= thr) & (gyro_var_2 <= thr) & (gyro_var_3 <= thr)
+        
+        # Exclude the last 'sampling_freq_Hz' samples from the identified periods
+        self.locs[-sampling_freq_Hz:] = False
+        
+        # If no bias periods are found, return the original gyro values
+        if np.sum(self.locs) == 0:
+            gyro_out = gyro
+            return gyro_out
+        
+        # Find the start and end indices of the first bias period
+        locs_place = np.where(self.locs)[0]
+        location = np.array([[locs_place[0]], [locs_place[0] + sampling_freq_Hz - 1]])
+        
+        # Calculate the bias for each axis within the identified period
+        self.gyro_bias = np.mean(gyro[location[0, 0]:location[1, 0] + 1], axis=0)
+        
+        # Subtract the bias from the original gyro signal
+        gyro = gyro - self.gyro_bias
+
+        # Integrate x component of the gyro signal to get yaw angle (also convert gyro unit to deg/s)
+        self.yaw = scipy.integrate.cumtrapz(gyro[:, 0] / self.gyro_convert_unit, initial=0) / sampling_freq_Hz
```

Comment:
Why do you convert the gyroscope units again, here?

Reply:
I re-organized unit conversion in whole code and now it is more consistent. Here, for calculation of yaw angles in degree I convert only once from rad/s to deg/s and the rest of the code continues using the rad/s data of gyroscope.